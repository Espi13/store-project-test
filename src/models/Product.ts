class Product {
  id: number;
  product_name: string;
  price: number;
  discount: number;
  image: string;

  constructor(id: number, product_name: string, price: number, discount: number, image: string) {
    this.id = id;
    this.product_name = product_name;
    this.price = price;
    this.discount = discount;
    this.image = image;
  }
}
export default Product;
