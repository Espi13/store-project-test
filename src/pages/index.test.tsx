import { fireEvent, render, screen } from '@testing-library/react';
import Product from 'models/Product';
import ProductPage, { getStaticProps } from './';

const mockProducts: Product[] = [
  {
    id: 1,
    product_name: 'Product 1',
    price: 10,
    discount: 5,
    image: 'http://dummyimage.com/3315x3575.png/cc0000/ffffff',
  },
  {
    id: 2,
    product_name: 'Product 2',
    price: 20,
    discount: null,
    image: 'http://dummyimage.com/3315x3575.png/cc0000/ffffff',
  },
  {
    id: 3,
    product_name: 'Product 3',
    price: 30,
    discount: null,
    image: 'http://dummyimage.com/3315x3575.png/cc0000/ffffff',
  },
];

describe('Start Product page', () => {
  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(mockProducts),
      }),
    );
  });

  it('getStaticProps return the correct list of products', async () => {
    const { props } = await getStaticProps();
    expect(props.products).toEqual(mockProducts);
  });

  it('renders searchInput component', async () => {
    const { container } = render(<ProductPage products={[]} />);
    const searchInput = screen.getByRole('textbox');
    expect(searchInput).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('filters the products when searching for a term', async () => {
    render(<ProductPage products={mockProducts} />);
    const searchInput = screen.getByRole('textbox');
    fireEvent.change(searchInput, { target: { value: 'Product 1' } });
    const product1 = screen.getByText('Product 1');
    const product2 = screen.queryByText('Product 2');
    const product3 = screen.queryByText('Product 3');
    expect(product1).toBeInTheDocument();
    expect(product2).not.toBeInTheDocument();
    expect(product3).not.toBeInTheDocument();
  });

  it('renders the ItemsRowChanger component', async () => {
    render(<ProductPage products={[]} />);
    const itemRowRemoveButton = screen.getByRole('button', { name: /remove/i });
    const itemRowAddButton = screen.getByRole('button', { name: /add/i });

    expect(itemRowRemoveButton).toBeInTheDocument();
    expect(itemRowAddButton).toBeInTheDocument();
  });

  it('changes the number of columns per item when selecting a different value', async () => {
    render(<ProductPage products={mockProducts} />);
    const itemRowRemoveButton = screen.getByRole('button', { name: /remove/i });
    fireEvent.click(itemRowRemoveButton);
    const productsContainer = screen.queryAllByTestId('product-item');
    expect(productsContainer).toHaveLength(3);
    for (const product of productsContainer) {
      expect(product).toHaveClass('MuiGrid-grid-md-4');
    }
  });
});
