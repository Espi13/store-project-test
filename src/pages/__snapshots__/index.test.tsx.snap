// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Start Product page renders searchInput component 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1vq5phj-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-11ndx0o-MuiGrid-root"
      >
        <div
          class="MuiGrid-root css-vj1n65-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1sh9ek1-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="SearchIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </div>
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
                id="search-product-input"
                placeholder="Buscar..."
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-ssy72w-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container css-8n36n9-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <button
              aria-label="remove"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
                data-testid="RemoveIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13H5v-2h14v2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <button
              aria-label="add"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-1f6u4kq-MuiGrid-root"
      data-testid="products"
    />
  </div>
</div>
`;
